proof/refine/RISCV64/Invariants_H.thy

```diff
@@ -169,50 +328,117 @@ definition ntfn_q_refs_of' :: "Structures_H.ntfn \<Rightarrow> (obj_ref \<times>
 definition ntfn_bound_refs' :: "obj_ref option \<Rightarrow> (obj_ref \<times> reftype) set" where
   "ntfn_bound_refs' t \<equiv> set_option t \<times> {NTFNBound}"
 
-definition tcb_bound_refs' :: "obj_ref option \<Rightarrow> (obj_ref \<times> reftype) set" where
-  "tcb_bound_refs' a \<equiv> set_option a \<times> {TCBBound}"
+definition refs_of_ntfn' :: "notification \<Rightarrow> ref_set" where
+  "refs_of_ntfn' ntfn \<equiv> ntfn_q_refs_of' (ntfnObj ntfn)
+                          \<union> get_refs NTFNBound (ntfnBoundTCB ntfn)
+                          \<union> get_refs NTFNSchedContext (ntfnSc ntfn)"
+
+definition refs_of_sc' :: "sched_context \<Rightarrow> ref_set" where
+  "refs_of_sc' sc \<equiv> get_refs SCNtfn (scNtfn sc)
+                          \<union> get_refs SCTcb (scTCB sc)
+                          \<union> get_refs SCYieldFrom (scYieldFrom sc)
+                          \<union> get_refs SCReply (scReply sc)"
+
+definition refs_of_reply' :: "reply \<Rightarrow> ref_set" where
+  "refs_of_reply' r \<equiv> get_refs ReplySchedContext (replySC r)
+                          \<union> get_refs ReplyTCB (replyTCB r)"
+
+definition list_refs_of_reply' :: "reply \<Rightarrow> ref_set" where
+  "list_refs_of_reply' r = get_refs ReplyNext (replyNext_of r) \<union> get_refs ReplyPrev (replyPrev r)"
+
+abbreviation list_refs_of_replies_opt' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> ref_set option" where
+  "list_refs_of_replies_opt' s \<equiv> replies_of' s ||> list_refs_of_reply'"
+
+(* FIXME RT: move to Option_Monad? *)
+definition map_set :: "('a \<Rightarrow> 'b set option) \<Rightarrow> 'a \<Rightarrow> 'b set" where
+  "map_set f \<equiv> case_option {} id \<circ> f"
```

Comment:
It's probably a good time to do this move now before we copy/paste it to more architectures in the future.